src/game/LFG/LFGMgr.h

```diff
@@ -0,0 +1,138 @@
+#include <map>
+
+#include "Policies/Singleton.h"
+#include "Common.h"
+
+enum ClassRoles
+{
+    LFG_ROLE_NONE   = 0x00,
+    LFG_ROLE_TANK   = 0x01,
+    LFG_ROLE_HEALER = 0x02,
+    LFG_ROLE_DPS    = 0x04
+};
+
+enum RolesPriority
+{
+    LFG_PRIORITY_NONE   = 0,
+    LFG_PRIORITY_LOW    = 1,
+    LFG_PRIORITY_NORMAL = 2,
+    LFG_PRIORITY_HIGH   = 3
+};
+
+enum PlayerLeaveMethod
+{
+    PLAYER_CLIENT_LEAVE = 0,
+    PLAYER_SYSTEM_LEAVE = 1
+};
+
+enum GroupLeaveMethod
+{
+    GROUP_CLIENT_LEAVE  = 0,
+    GROUP_SYSTEM_LEAVE  = 1
+};
+
+#define MAX_DPS_COUNT = 3
+
+struct LFGPlayerQueueInfo
+{
+    ClassRoles roleMask;
+    uint32 team;
+    uint32 areaId;
+    uint32 timeInLFG;
+    bool hasQueuePriority;
+    std::string name;
+    std::list<std::pair<ClassRoles, RolesPriority>> rolePriority;
+
+    void CalculateRoles(Classes playerClass);
+    void CalculateTalentRoles(Player* player);
+    RolesPriority GetRolePriority(ClassRoles role);
+};
+
+struct LFGGroupQueueInfo
+{
+    uint32 availableRoles;
+    uint32 dpsCount;
+    uint32 team;
+    uint32 areaId;
+    uint32 groupTimer;
+};
+
+struct MeetingStoneInfo
+{
+    uint32 area;
+    uint32 minlevel;
+    uint32 maxlevel;
+    char*  name;
+};
+
+typedef std::vector<MeetingStoneInfo> MeetingStoneSet;
+
+class LFGQueue
+{
+    public:
+        LFGQueue() {}
+        ~LFGQueue() {}
+
+        void AddToQueue(Player* leader, uint32 queAreaID);
+        void RestoreOfflinePlayer(Player* player);
+        bool IsPlayerInQueue(ObjectGuid const& plrGuid) const;
+        bool IsGroupInQueue(uint32 groupId) const;
+        void RemovePlayerFromQueue(ObjectGuid const& plrGuid, PlayerLeaveMethod leaveMethod = PLAYER_CLIENT_LEAVE); // 0 == by default system (cmsg, leader leave), 1 == by lfg system (no need report text you left queu)
+        void RemoveGroupFromQueue(uint32 groupId, GroupLeaveMethod leaveMethod = GROUP_CLIENT_LEAVE);
+        void Update(uint32 diff);
+        void UpdateGroup(uint32 groupId);
+        void GetPlayerQueueInfo(LFGPlayerQueueInfo* info, ObjectGuid const& plrGuid);
+        void GetGroupQueueInfo(LFGGroupQueueInfo* info, uint32 groupId);
+
+        static void BuildSetQueuePacket(WorldPacket& data, uint32 areaId, uint8 status);
+        static void BuildMemberAddedPacket(WorldPacket& data, ObjectGuid plrGuid);
+        static void BuildInProgressPacket(WorldPacket& data);
+        static void BuildCompletePacket(WorldPacket& data);
+
+        static ClassRoles CalculateRoles(Classes playerClass);
+        static ClassRoles CalculateTalentRoles(Player* player);
+        static int GetHighestTalentTree(Player* player);
+        static std::map<uint32, int32> GetTalentTrees(Player* player);
+        static RolesPriority getPriority(Classes playerClass, ClassRoles playerRoles);
+
+        static uint32 GetMaximumDPSSlots() { return 3u; }
+
+    private:
+        typedef std::map<ObjectGuid, LFGPlayerQueueInfo> QueuedPlayersMap;
+        QueuedPlayersMap m_QueuedPlayers;
```

Comment:
cmangos uses codestyle of m_lowerCase in the newer code (still trying to unify the old code slowly)